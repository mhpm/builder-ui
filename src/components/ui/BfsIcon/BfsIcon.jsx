import React from 'react';
import {
  FaSpinner,
  FaCog,
  FaCircleNotch,
  FaStar,
  FaRegStar,
} from 'react-icons/fa';
import PropTypes from 'prop-types';

import { css } from '@emotion/react';

const stIcon = css`
  margin: 0;
  padding: 0;
`;

const stSpin = css`
  animation: spin 2s infinite linear;
`;

export const icons = {
  star: FaStar,
  starhollow: FaRegStar,
  spinner: FaSpinner,
  cog: FaCog,
  circlenotch: FaCircleNotch,
};

/**
 * Reusable icon component that follows the conventions from the BFS Design and
 * Component Library.
 *
 * It comes in 3 different icons to chose using the `icon` property:
 * ***spinner***, ***cog***, and ***circlenotch***.
 *
 * The icon can be set to spin so we can use it as spinner by the `spin`
 * boolean property.
 *
 * The size can be set by the `size` property.
 *
 */
export const BfsIcon = ({
  'data-testid': dataTestId,
  icon,
  spin,
  size,
}) => {
  const IconComponent = icons[icon];

  return (
    <IconComponent
      data-testid={dataTestId}
      css={[stIcon, spin && stSpin]}
      size={size}
    />
  );
};

BfsIcon.propTypes = {
  /**
   * Test id only for unit testing element selection.
   */
  'data-testid': PropTypes.string,

  /**
   * The name of the icon to show.
   */
  icon: PropTypes.oneOf(Object.keys(icons)),

  /**
   * Flag to apply spinning animation
   */
  spin: PropTypes.bool,

  /**
   * Icon size in pixels
   */
  size: PropTypes.string,
};

BfsIcon.displayName = 'BfsIcon';

BfsIcon.defaultProps = {
  'data-testid': undefined,
  icon: 'spinner',
  spin: false,
  size: '16',
};
