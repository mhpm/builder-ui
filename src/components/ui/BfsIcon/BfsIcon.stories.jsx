/* eslint-disable react/jsx-props-no-spreading */
import { BfsIcon, icons } from './BfsIcon';

export default {
  title: 'Atoms/BfsIcon',
  component: BfsIcon,
  parameters: {
    componentSubtitle: 'Reusable Icons for BFS projects',
  },
  argTypes: {
    'data-testid': {
      control: 'none',
    },
    icon: {
      control: {
        type: 'select',
        options: Object.keys(icons),
      },
    },
  },
  args: {
    'data-testid': '',
    icon: 'spinner',
    spin: false,
    size: '16',
  },
};

export const Default = {
  render: (args) => (
    <BfsIcon {...args} />
  ),
};

export const Spin = {
  args: {
    icon: 'spinner',
    spin: true,
  },
  render: (args) => <BfsIcon {...args} />,
};

export const Gallery = {
  args: {
  },
  render: (args) => (
    <div style={{ display: 'flex', gap: 10 }}>
      {Object.keys(icons).map((icon) => (
        <BfsIcon key={icon} {...args} icon={icon} />
      ))}
    </div>
  ),
};
