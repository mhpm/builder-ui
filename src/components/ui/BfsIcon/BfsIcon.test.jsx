import React from 'react';
import { render } from '@testing-library/react';
import { matchers } from '@emotion/jest';

import { BfsIcon, icons } from './BfsIcon';

expect.extend(matchers);

describe('BfsIcon component render', () => {
  it('correctly without crashing with testid and the spinner icon by default', () => {
    const { getByTestId } = render(<BfsIcon data-testid="icon" />);
    const icon = getByTestId('icon');

    expect(icon).toBeInTheDocument();
    expect(icon).toMatchSnapshot();
  });
});

describe('BfsIcon component property', () => {
  Object.keys(icons).forEach((icon) => {
    it(`icon ${icon}`, () => {
      const { getByTestId } = render(<BfsIcon data-testid="icon" icon={icon} />);
      const iconCmp = getByTestId('icon');

      expect(iconCmp).toBeInTheDocument();
      expect(iconCmp).toMatchSnapshot();
    });
  });

  test('applies custom className when provided', () => {
    const { getByTestId } = render(<BfsIcon data-testid="icon" />);

    expect(getByTestId('icon')).toBeInTheDocument();
  });
});
