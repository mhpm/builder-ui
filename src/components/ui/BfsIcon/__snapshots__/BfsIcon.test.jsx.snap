// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BfsIcon component property icon circlenotch 1`] = `
.emotion-0 {
  margin: 0;
  padding: 0;
}

<svg
  class="emotion-0"
  data-testid="icon"
  fill="currentColor"
  height="16"
  stroke="currentColor"
  stroke-width="0"
  viewBox="0 0 512 512"
  width="16"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M288 39.056v16.659c0 10.804 7.281 20.159 17.686 23.066C383.204 100.434 440 171.518 440 256c0 101.689-82.295 184-184 184-101.689 0-184-82.295-184-184 0-84.47 56.786-155.564 134.312-177.219C216.719 75.874 224 66.517 224 55.712V39.064c0-15.709-14.834-27.153-30.046-23.234C86.603 43.482 7.394 141.206 8.003 257.332c.72 137.052 111.477 246.956 248.531 246.667C393.255 503.711 504 392.788 504 256c0-115.633-79.14-212.779-186.211-240.236C302.678 11.889 288 23.456 288 39.056z"
  />
</svg>
`;

exports[`BfsIcon component property icon cog 1`] = `
.emotion-0 {
  margin: 0;
  padding: 0;
}

<svg
  class="emotion-0"
  data-testid="icon"
  fill="currentColor"
  height="16"
  stroke="currentColor"
  stroke-width="0"
  viewBox="0 0 512 512"
  width="16"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M487.4 315.7l-42.6-24.6c4.3-23.2 4.3-47 0-70.2l42.6-24.6c4.9-2.8 7.1-8.6 5.5-14-11.1-35.6-30-67.8-54.7-94.6-3.8-4.1-10-5.1-14.8-2.3L380.8 110c-17.9-15.4-38.5-27.3-60.8-35.1V25.8c0-5.6-3.9-10.5-9.4-11.7-36.7-8.2-74.3-7.8-109.2 0-5.5 1.2-9.4 6.1-9.4 11.7V75c-22.2 7.9-42.8 19.8-60.8 35.1L88.7 85.5c-4.9-2.8-11-1.9-14.8 2.3-24.7 26.7-43.6 58.9-54.7 94.6-1.7 5.4.6 11.2 5.5 14L67.3 221c-4.3 23.2-4.3 47 0 70.2l-42.6 24.6c-4.9 2.8-7.1 8.6-5.5 14 11.1 35.6 30 67.8 54.7 94.6 3.8 4.1 10 5.1 14.8 2.3l42.6-24.6c17.9 15.4 38.5 27.3 60.8 35.1v49.2c0 5.6 3.9 10.5 9.4 11.7 36.7 8.2 74.3 7.8 109.2 0 5.5-1.2 9.4-6.1 9.4-11.7v-49.2c22.2-7.9 42.8-19.8 60.8-35.1l42.6 24.6c4.9 2.8 11 1.9 14.8-2.3 24.7-26.7 43.6-58.9 54.7-94.6 1.5-5.5-.7-11.3-5.6-14.1zM256 336c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z"
  />
</svg>
`;

exports[`BfsIcon component property icon spinner 1`] = `
.emotion-0 {
  margin: 0;
  padding: 0;
}

<svg
  class="emotion-0"
  data-testid="icon"
  fill="currentColor"
  height="16"
  stroke="currentColor"
  stroke-width="0"
  viewBox="0 0 512 512"
  width="16"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
  />
</svg>
`;

exports[`BfsIcon component property icon star 1`] = `
.emotion-0 {
  margin: 0;
  padding: 0;
}

<svg
  class="emotion-0"
  data-testid="icon"
  fill="currentColor"
  height="16"
  stroke="currentColor"
  stroke-width="0"
  viewBox="0 0 576 512"
  width="16"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
  />
</svg>
`;

exports[`BfsIcon component property icon starhollow 1`] = `
.emotion-0 {
  margin: 0;
  padding: 0;
}

<svg
  class="emotion-0"
  data-testid="icon"
  fill="currentColor"
  height="16"
  stroke="currentColor"
  stroke-width="0"
  viewBox="0 0 576 512"
  width="16"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z"
  />
</svg>
`;

exports[`BfsIcon component render correctly without crashing with testid and the spinner icon by default 1`] = `
.emotion-0 {
  margin: 0;
  padding: 0;
}

<svg
  class="emotion-0"
  data-testid="icon"
  fill="currentColor"
  height="16"
  stroke="currentColor"
  stroke-width="0"
  viewBox="0 0 512 512"
  width="16"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
  />
</svg>
`;
